### PR TITLE
[feat][refactor][OSS] Param buckets + fp16 broadcasts

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Similar to the grad buckets but with a different feature set (and small enough so that I felt that it was worth a minor duplication), add a class to represent the bucketed parameters, and make it easy to compress them to fp16 and subsequently decompress. ~20% speed boost with slow interconnect, no real accuracy loss with AMP (the FW is done with fp16 parameters anyway). Useless with fairseq for instance (parameters are already fp16 to begin with)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.
cc @mannatsingh, a proper implementation of what talked about a couple of months back

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/misc/grad_bucket.py [hunk 1]
```diff
@@ -16,6 +16,7 @@ class GradBucket:
     def __init__(self, size: int, dtype: torch.dtype, device: torch.device, destination: int) -> None:
         self._max_size = size
         self._params: List[torch.Tensor] = []
+        self._param_ids: List[int] = []
```
Comment (blefaudeux) [inline fairscale/nn/misc/grad_bucket.py:19]: small extension to fool proof these: forbid the same params being checked in twice

File: fairscale/nn/misc/grad_bucket.py [hunk 2]
```diff
@@ -70,11 +71,15 @@ def add_grad(self, param: torch.Tensor) -> None:
         """
         Add a new parameter gradient to the bucket. Param.grad becomes a view of this bucket buffer
         """
+
+        assert id(param) not in self._param_ids, "The same gradients cannot be checked in twice"
```
Comment (min-xu-ai) [inline fairscale/nn/misc/grad_bucket.py:75]: when a module is forwarded multiple times, like the multi_res_input_forward in vissl, perhaps this will happen?

File: fairscale/optim/oss.py [hunk 3]
```diff
@@ -52,6 +54,10 @@ class OSS(Optimizer):
             torch.distributed group (default: group.WORLD)
         broadcast_buffer_size (int):
             (deprecated) used to cap the size of the broadcast buffers, not being used anymore.
+        broadcast_fp16 (bool):
+            Compress the model shards in fp16 before sharing them in between ranks.
```
Comment (min-xu-ai) [inline fairscale/optim/oss.py:58]: model shards or optimizer state shards?

**Review by min-xu-ai — COMMENTED** [review]
I am wondering can grad bucket and param bucket share something? Should they have separate classes? To me, they are both tensors and both bucketing "horizontally", i.e. between params. Is there some opportunity to simplify the code here?

File: fairscale/nn/misc/grad_bucket.py (hunk 1 above)
Comment (anj-s) [reply fairscale/nn/misc/grad_bucket.py:19]: Just curious if you have seen this happen. Can you have grads checked in twice?

File: fairscale/optim/oss.py (hunk 3 above)
Comment (blefaudeux) [reply fairscale/optim/oss.py:58]: model shards, really, not like in FDSP since each rank has all of them, but these are real model shards

File: fairscale/nn/misc/grad_bucket.py (hunk 2 above)
Comment (blefaudeux) [reply fairscale/nn/misc/grad_bucket.py:75]: it's defined at construction time, not per forward, so this should be fine. I think that the assert is important because the grad becomes a view of a tensor, and it cannot be a (correct) view of two things at the same time

File: fairscale/nn/misc/param_bucket.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Union
+
+import torch
+
+
+class ParamBucket:
```
Comment (anj-s) [inline fairscale/nn/misc/param_bucket.py:11]: Why not just a Bucket class? Seem pretty similar to GradBuckets?

**Comment by blefaudeux** [comment]
> I am wondering can grad bucket and param bucket share something? Should they have separate classes? To me, they are both tensors and both bucketing "horizontally", i.e. between params. Is there some opportunity to simplify the code here?

I was wondering about that, the shared part is that there's an underlying flat buffer and that tensors are checked in. This being said they also have significant differences (grads can be None, not params, when checking in in the grad bucket the param is the reference but the grad data is checked in, vs. param all the way for the param bucket, collapsing/rebuilding is only needed for grad bucket, etc..). I think that both could inherit a base class and this would remove ~10 lines of code, not sure whether that's worth it but could be done in another PR if you'd like

File: fairscale/nn/misc/param_bucket.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        assert param.dtype == self.buffer.dtype
+        assert param.device == self.buffer.device
+
+        fill_next = self._fill + param.numel()
+        assert fill_next <= self.buffer.numel()
+
+        # Copy the current param value
+        if keep_existing_value:
+            self.buffer[self._fill : fill_next].copy_(param.data.flatten())
+        param.data = self.buffer[self._fill : fill_next].view_as(param.data)
```
Comment (anj-s) [inline fairscale/nn/misc/param_bucket.py:65]: why do you need to do this?

File: fairscale/nn/misc/param_bucket.py (hunk 4 above)
Comment (blefaudeux) [reply fairscale/nn/misc/param_bucket.py:11]: see https://github.com/facebookresearch/fairscale/pull/540#issuecomment-803023686, there are differences still, I can make them inherit from the same base class if you'd like but don't expect huge line count savings. I'm being told that my PRs are too big, so probably better to keep that for another one ?

File: fairscale/optim/oss.py [hunk 6]
```diff
@@ -52,6 +54,10 @@ class OSS(Optimizer):
...
+            This is safe to use when PyTorch AMP is activated. Without torch AMP this will lead to a slight
+            degradation in terms of accuracy.
```
Comment (anj-s) [inline fairscale/optim/oss.py:60]: Should we disallow it in that case?

File: fairscale/nn/misc/param_bucket.py (hunk 5 above)
Comment (blefaudeux) [reply fairscale/nn/misc/param_bucket.py:65]: "this" being ? this precise line is the tensor view

File: fairscale/nn/misc/grad_bucket.py (hunk 1 above)
Comment (blefaudeux) [reply fairscale/nn/misc/grad_bucket.py:19]: never say never :) Some models share weights in different places, so yes I think this could easily happen, I was trying to be cautious

File: tests/optim/test_oss.py [hunk 7]
```diff
@@ -501,7 +501,7 @@ def run_test_reproducibility(rank, world_size, tempfile_name):
     loss_fn = torch.nn.L1Loss()
     loss_fn.to(device)
 
-    optimizer = optim.OSS(model.parameters(), optim=torch.optim.RMSprop, lr=0.1)
+    optimizer = optim.OSS(model.parameters(), optim=torch.optim.RMSprop, lr=0.1, broadcast_fp16=broadcast_fp16)
```
Comment (anj-s) [inline tests/optim/test_oss.py:504]: Can we do a weight check as well? I think you have a test for that as well if i remember correctly.

File: tests/optim/test_oss.py [hunk 8]
```diff
@@ -534,12 +534,13 @@ def closure():
...
 @skip_if_single_gpu
-def test_reproducibility():
+@pytest.mark.parametrize("broadcast_fp16", [False, True])
+def test_reproducibility(broadcast_fp16: bool):
     world_size = 2
     temp_file_name = tempfile.mkstemp()[1]
 
     mp.spawn(
-        run_test_reproducibility, args=(world_size, temp_file_name), nprocs=world_size, join=True,
+        run_test_reproducibility, args=(world_size, temp_file_name, broadcast_fp16), nprocs=world_size, join=True,
```
Comment (anj-s) [inline tests/optim/test_oss.py:543]: Shall we add a benchmark for this? WDYT?

File: tests/optim/test_oss.py (hunk 7 above)
Comment (blefaudeux) [reply tests/optim/test_oss.py:504]: the bucket test is doing that, here I think that the comparison would be not completely trivial, because some accuracy is lost when doing fp32->fp16->fp32, so the result will not be strictly the same as not doing it. The bucket unit test does check that the results are the same as doing this conversion chain outside of the bucket

File: tests/optim/test_oss.py (hunk 8 above)
Comment (blefaudeux) [reply tests/optim/test_oss.py:543]: ideally it would be on fb cluster, it only really makes sense when inter node

**Comment by blefaudeux** [comment]
@min-xu-ai @anj-s I can make both bucket inherit from a base class, it's true that they share some things (and for instance the type change would be handy for fp16 reduce support with buckets for ShardedDDP), this PR or another one ?

File: fairscale/nn/misc/param_bucket.py (hunk 4 above)
Comment (anj-s) [reply fairscale/nn/misc/param_bucket.py:11]: I was only thinking in terms of making changes in two places. If you end up doing that you can rethink merging this. For now this looks great! very clean and easy to understand.

File: fairscale/nn/misc/param_bucket.py (hunk 5 above)
Comment (anj-s) [reply fairscale/nn/misc/param_bucket.py:65]: From offline discussion: The params and buckets share the same storage but have different views.

File: tests/optim/test_oss.py (hunk 7 above)
Comment (anj-s) [reply tests/optim/test_oss.py:504]: discussed offline: Since the weights may not be exactly equal we want to know what the tolerance diff value could be.

File: tests/optim/test_oss.py (hunk 8 above)
Comment (anj-s) [reply tests/optim/test_oss.py:543]: from offline discussion since I was not clear why this only makes sense for inter node: Turns our the perf gain is insignificant due to fast intranode comms. We still want to measure performance to make sure there are no regressions.

File: fairscale/optim/oss.py (hunk 6 above)
Comment (anj-s) [reply fairscale/optim/oss.py:60]: Makes sense. Might be good to document the accuracy loss that you came across in your testing.

**Comment by anj-s** [comment]
> @min-xu-ai @anj-s I can make both bucket inherit from a base class, it's true that they share some things (and for instance the type change would be handy for fp16 reduce support with buckets for ShardedDDP), this PR or another one ?

I always vote for new PRs so that we can quickly close and iterate on this one :) Upto you though, whatever you feel is best.

**Review by anj-s — APPROVED** [review]

**Comment by blefaudeux** [comment]
@min-xu-ai @anj-s unified buckets (well, through inheritance), still ok for you ?

File: tests/optim/test_oss.py (hunk 8 above)
Comment (blefaudeux) [reply tests/optim/test_oss.py:543]: I can add that in another PR if that's ok ? I tested locally and not a big difference, if anything a bit slower with this option turned on, as expected

**Comment by blefaudeux** [comment]
> @min-xu-ai @anj-s unified buckets (well, through inheritance), still ok for you ?

guess is yes :)
